libsmtutil/SMTLib2Interface.cpp

```diff
@@ -212,33 +213,40 @@ std::string SMTLib2Interface::toSExpr(Expression const& _expr)
 	return sexpr;
 }
 
-std::string SMTLib2Interface::toSmtLibSort(Sort const& _sort)
+std::string SMTLib2Interface::toSmtLibSort(solidity::smtutil::SortPointer _sort)
 {
-	switch (_sort.kind)
+	if (!m_sortNames.count(_sort))
+		m_sortNames[_sort] = toSmtLibSortInternal(_sort);
+	return m_sortNames.at(_sort);
+}
+
+std::string SMTLib2Interface::toSmtLibSortInternal(SortPointer _sort)
+{
+	switch (_sort->kind)
 	{
 	case Kind::Int:
 		return "Int";
 	case Kind::Bool:
 		return "Bool";
 	case Kind::BitVector:
-		return "(_ BitVec " + std::to_string(dynamic_cast<BitVectorSort const&>(_sort).size) + ")";
+		return "(_ BitVec " + std::to_string(dynamic_cast<BitVectorSort const&>(*_sort).size) + ")";
 	case Kind::Array:
 	{
-		auto const& arraySort = dynamic_cast<ArraySort const&>(_sort);
+		auto const& arraySort = dynamic_cast<ArraySort const&>(*_sort);
 		smtAssert(arraySort.domain && arraySort.range, "");
-		return "(Array " + toSmtLibSort(*arraySort.domain) + ' ' + toSmtLibSort(*arraySort.range) + ')';
+		return "(Array " + toSmtLibSort(arraySort.domain) + ' ' + toSmtLibSort(arraySort.range) + ')';
 	}
 	case Kind::Tuple:
 	{
-		auto const& tupleSort = dynamic_cast<TupleSort const&>(_sort);
+		auto const& tupleSort = dynamic_cast<TupleSort const&>(*_sort);
 		std::string tupleName = "|" + tupleSort.name + "|";
 		auto isName = [&](auto entry) { return entry.first == tupleName; };
 		if (ranges::find_if(m_userSorts, isName) == m_userSorts.end())
 		{
 			std::string decl("(declare-datatypes ((" + tupleName + " 0)) (((" + tupleName);
 			smtAssert(tupleSort.members.size() == tupleSort.components.size(), "");
 			for (unsigned i = 0; i < tupleSort.members.size(); ++i)
-				decl += " (|" + tupleSort.members.at(i) + "| " + toSmtLibSort(*tupleSort.components.at(i)) + ")";
+				decl += " (|" + tupleSort.members.at(i) + "| " + toSmtLibSort(tupleSort.components.at(i)) + ")";
```

Comment:
I know that's not you but I would switch lines 246 and 247, so `decl` is declared closer to it's use.

Reply:
I agree, but I prefer not to touch that code now.